File: src/main/java/artisra/topinterview150/hashmap/LongestConsecutiveSequence.java
package artisra.topinterview150.hashmap;

import java.util.HashSet;
import java.util.Set;

/**
 * This class provides a solution to the problem of finding the longest consecutive sequence in an unsorted array of integers.
 * It implements an efficient algorithm with a time complexity of O(N) and a space complexity of O(N).
 * @see <a href="https://leetcode.com/problems/longest-consecutive-sequence/solutions/5987937/java-kotlin-time-o-n-space-o-n">My Solution on LeetCode.</a>
 */
public class LongestConsecutiveSequence {
    /**
     * Finds the length of the longest consecutive sequence in the given array of integers.
     *
     * @param nums the input array of integers
     * @return the length of the longest consecutive sequence
     */
    public int longestConsecutive(int[] nums) {

        // Storing all the numbers ina hashset.
        Set<Integer> availableNumbers = new HashSet<>() {{
            for (int num : nums) add(num);
        }};

        int longestSeqSize = 0;

        for (int num : nums) {
            if (availableNumbers.isEmpty()) break; // No more items to process.
            int currentNumber = num;
            if (!availableNumbers.contains(currentNumber))
                continue; // This number and its sequence have been processed.

            int seqSize = 1;

            while (availableNumbers.contains(--currentNumber)) {
                seqSize++;
                availableNumbers.remove(currentNumber);
            }

            currentNumber = num;
            while (availableNumbers.contains(++currentNumber)) {
                seqSize++;
                availableNumbers.remove(currentNumber);
            }

            availableNumbers.remove(num);
            if (seqSize > longestSeqSize) longestSeqSize = seqSize;
        }

        return longestSeqSize;
    }
}

File: src/main/kotlin/topinterview150/hashmap/LongestConsecutiveSequence.kt
package dev.artisra.topinterview150.hashmap

/**
 * This class provides a solution to the problem of finding the longest consecutive sequence in an unsorted array of integers.
 * It implements an efficient algorithm with a time complexity of O(N) and a space complexity of O(N).
 * @see <a href="https://leetcode.com/problems/longest-consecutive-sequence/solutions/5987937/java-kotlin-time-o-n-space-o-n">My Solution on LeetCode.</a>
 */
class LongestConsecutiveSequence {
    /**
     * Finds the length of the longest consecutive sequence in the given array of integers.
     *
     * @param nums the input array of integers
     * @return the length of the longest consecutive sequence
     */
    fun longestConsecutive(nums: IntArray): Int {
        val availableNumbers = nums.toMutableSet()
        var longestSeqSize = 0

        for (num in nums) {
            if (availableNumbers.isEmpty()) break // No more numbers to check

            var currentNumber = num

            if (currentNumber !in availableNumbers) continue // This number and its sequence has been processed.

            var currentSeqSize = 1

            // Checking for sequence downwards.
            while (--currentNumber in availableNumbers) {
                currentSeqSize++
                availableNumbers.remove(currentNumber)
            }
            currentNumber = num

            // Checking for sequence upwards.
            while (++currentNumber in availableNumbers) {
                currentSeqSize++
                availableNumbers.remove(currentNumber)
            }
            availableNumbers.remove(num)
            longestSeqSize = longestSeqSize.coerceAtLeast(currentSeqSize)
        }

        return longestSeqSize
    }
}
